pkg/kmodule/kmodule_linux.go

```diff
@@ -261,3 +257,29 @@ func loadModule(path, modParams string, opts ProbeOpts) error {
 
 	return nil
 }
+
+func genLoadedMods(deps depMap) error {
```

Comment:
u-root has historically had weak support for testing (this is my fault) but we are trying to fix that. If you could change this:
genLoadedMods(r io.Reader, deps depMap) error

then you could put the open of /proc/modules outside it.
then
scanner := bufio.NewScanner(r)

and then it would be pretty easy to write a test for this function, which seems pretty critical. 
Sound ok?

Reply:
Sounds good. Coding with testability in mind sounds like a good habit 

I would argue that this function should not be considered critical though. You don't want to fail in case that something goes wrong reading/parsing /proc/modules. In the worst case scenario (e.g. proc is not mounted) you will attempt to load a module that is already loaded. Not sure if that results in an error though.

So the calling implementation could be simplified to
```go
fm, err := os.Open("/proc/modules")
if err == nil {
	defer fm.Close()
	genLoadedMods(fm, deps)
}
```
What do you think?

Reply:
sure, I'm not sure the defer is strictly necessary:
getLoadedMods(fm, deps)
fm.Close()

I've gotten in the habit of avoiding close by defer since it has burned me in the past. Either, way, sounds fine.

Reply:
Hate to contradict, but please defer. Defer on close only burns you for writable files. And still burns you if you do it like this, because you didn't check the error. Which doesn't matter on read-only files.

Reply:
Hi @hugelgupf. 2 questions

Is there a case when os.Open returns an error and i still need to call f.Close()?

What is the difference between
```go
fm, err := os.Open("/proc/modules")
if err == nil {
	defer fm.Close()
	genLoadedMods(fm, deps)
}
```
and
```go
fm, err := os.Open("/proc/modules")
if err == nil {
	genLoadedMods(fm, deps)
	fm.Close()
}
```
besides the stylistic/consistency  aspect?

Other than that i see no issue reverting this to use defer.

Reply:
it's fine, use defer. I did get burned by it even on readonly files but this is not a case where it could occur.

Reply:
also if the err is non-nill then fm will be nil. So you can't do much with it anyway ... so this is fine, you only do the close if err == nil

Reply:
@rminnich defer it is then.

But can someone answer my questions? I would like to avoid shooting my self in the foot in the future so please share your experience :)

Reply:
> also if the err is non-nill then fm will be nil. So you can't do much with it anyway ... so this is fine, you only do the close if err == nil

Thank you for clarifying this, my thoughts exactly.